src/RequestParser.cpp

```diff
@@ -416,40 +418,48 @@ void RequestParser::parseHeaders(HTTPRequest& request)
  * parser.parseChunkedBody(requestStream);
  * @endcode
  */
-void RequestParser::parseChunkedBody(HTTPRequest& request)
+void RequestParser::parseChunkedBody(HTTPRequest& request, std::vector<char>& buffer)
 {
+    LOG_DEBUG << "Parsing chunked body...";
+
 	size_t length = 0;
 	std::string strChunkSize;
-	std::string chunkData;
 	size_t numChunkSize = 0;
 
 	do {
 		std::getline(m_requestStream, strChunkSize);
-		if (strChunkSize[strChunkSize.size() - 1] == '\r')
+		if (!strChunkSize.empty() && strChunkSize[strChunkSize.size() - 1] == '\r')
 			strChunkSize.erase(strChunkSize.size() - 1);
 		else {
 			request.httpStatus = StatusBadRequest;
 			request.shallCloseConnection = true;
 			throw std::runtime_error(ERR_MISS_CRLF);
 		}
 		numChunkSize = convertHex(strChunkSize);
-		std::getline(m_requestStream, chunkData);
-		if (chunkData[chunkData.size() - 1] == '\r')
-			chunkData.erase(chunkData.size() - 1);
-		else {
-			request.httpStatus = StatusBadRequest;
+        buffer.resize(numChunkSize + 2);
+
+		errno = 0;
+		m_requestStream.read(buffer.data(), static_cast<long>(numChunkSize + 2));
+		if (m_requestStream.gcount() != static_cast<std::streamsize>(numChunkSize + 2)) {
+			request.httpStatus = StatusInternalServerError;
```

Comment:
Added check for goodbit: if it fails, then it was indeed a read error, set StatusInternalServerError and throw errno.

Otherwise it was a wrong size indication, set StatusBadRequest and throw ERR_CHUNKSIZE_INCONSISTENT (renamed ERR_CHUNK_SIZE, it was too general).

Reply:
I was confused at first why eofbit but get it now: if nothing is left to read. I think in this case we could use .fail() since it checks if badbit or failbit are set (and not goodbit and eofbit): https://cplusplus.com/reference/ios/ios/fail/

Reply:
One would think at first sight that both _if (!m_requestStream.good() && !m_requestStream.eof()) _ and _if (m_request.fail())_ are equal. But actually it is not:

When you reach EOF while reading just so, then the eofbit is set and both conditions would equally return false. But if read() is instructed to retrieve more data than is actually available, then both the eofbit and the failbit are set. Then the first condition will return false whereas the second one will return true. And that is our case - or what we expect to catch with our ERR_CHUNKSIZE_INCONSISTENT. That a larger chunksize is indicated than available and we reach EOF and want to read past it, but can't.

What I'm trying to say is we only want that condition to be true, when an actual read() error occurs and not when an error happens due to false size indications. That is handled afterwards.

Reply:
I see, didn't know that!